### PR TITLE
ENH: extra vars to update.sh

### PR DESCRIPTION
**Pull request by nicovanniekerk**
allowing to pass extra_vars from update script to the ansible plays being called

Configuration Pull Request
---

Make sure that the following steps are done before merging:

  - [ ] A DevOps team member has approved the PR.
  - [ ] Are you adding any new default values that need to be overridden when this change goes live? If so:
    - [ ] Update the appropriate internal repo (be sure to update for all our environments)
    - [ ] If you are updating a secure value rather than an internal one, file a DEVOPS ticket with details.
    - [ ] Add an entry to the CHANGELOG.
  - [ ] Have you performed the proper testing specified on the [Ops Ansible Testing Checklist](https://openedx.atlassian.net/wiki/display/EdxOps/Ops+Ansible+Testing+Checklist)?

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @nicovanniekerk! I've created [OSPR-1796](https://openedx.atlassian.net/browse/OSPR-1796) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



We can't start reviewing your pull request until you've submitted a     [signed contributor agreement](http://open.edx.org/sites/default/files/wysiwyg/individual-contributor-agreement.pdf)     or indicated your institutional affiliation.     If you like, you can add yourself to the [AUTHORS](https://github.com/edx/configuration/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

**Comment by gsong** [comment]
@nicovanniekerk Thank you for your PR. We appreciate you taking the time to make the contribution.

Please submit a [signed contributor agreement](http://open.edx.org/sites/default/files/wysiwyg/individual-contributor-agreement.pdf) to us so we can review this pull request. We realize it's additional work for a simple change in this case, but you only have to do this *once* and we're hoping you would make future contributions as well.

**Comment by nicovanniekerk** [comment]
Hi George,

Thanks for checking in, i am just waiting for the signed entity agreement,
as soon as i have it i will forward it to you.

Thank you

**Comment by nicovanniekerk** [comment]
Hi George,

Please find attached the completed entity contribution agreement.

Kind regards

File: playbooks/roles/edx_ansible/templates/update.j2 [hunk 1]
```diff
@@ -44,6 +44,10 @@ fi
 extra_args="$extra_args -e 'disable_edx_services=true' -e devstack=true"
 {% endif %}
 
+if [[ "$3" == "-e" ]]; then
+        extra_args="$extra_args -e ${@:3}"
```
Comment (feanil) [inline playbooks/roles/edx_ansible/templates/update.j2]: What does `${@:3}` do?  I haven't seen that expansion before.  Can you link to some docs?

Comment (nicovanniekerk) [reply playbooks/roles/edx_ansible/templates/update.j2]: Hi, @feanil please see this link[http://wiki.bash-hackers.org/scripting/posparams#mass_usage] for examples and explanation. 

Simply put, that expression returns all passed command line arguments after the 3rd position in the list.

Comment (feanil) [reply playbooks/roles/edx_ansible/templates/update.j2]: Gotcha makes sense.  I'm not sure I like it taking all the rest of the command line arguments, this makes it really hard to add any other options in the future without breaking people.  What do you think about just using args 3 and 4. If you need to make multiple overrides you can just point to a YAML file.

Comment (nicovanniekerk) [reply playbooks/roles/edx_ansible/templates/update.j2]: It does make sense, yes. I will change it to have the 3rd argument as the switch and the 4th will hold all extra vars as a string or as a yaml file.

**Comment by nicovanniekerk** [comment]
@feanil please see applied changes

**Comment by nicovanniekerk** [comment]
@pomegranited @feanil please see applied changes

**Comment by pomegranited** [comment]
Hi @nicovanniekerk!  I'm not sure why you pinged me on this PR..  I haven't worked on this script before?
@feanil did you want me to review it, or is there someone on your team that's better equipped to give a second review?

**Comment by gsong** [comment]
@nicovanniekerk I don't see an attachment for the CLA. Please email it to legal@edx.org per instructions at the bottom of the form. Thank you.

**Comment by nicovanniekerk** [comment]
Hi George, i have submitted the entity agreement about 2 weeks ago. please
ask the legal guys if the received it, if not i will send it again.

Thank you

File: playbooks/roles/edx_ansible/templates/update.j2 [hunk 2]
```diff
@@ -44,6 +48,12 @@ fi
 extra_args="$extra_args -e 'disable_edx_services=true' -e devstack=true"
 {% endif %}
 
+if [[ "$3" == "-e" ]]; then
+        opt=$4
+        sed -e 's/^"//' -e 's/"$//' <<<"$opt"
```
Comment (nedbat) [inline playbooks/roles/edx_ansible/templates/update.j2]: Seems like this line is meant to remove double quotes from opt, but does it put the results back into opt for use in the next line?

**Comment by gsong** [comment]
@nicovanniekerk I've sent a follow up email regarding the CLA, we're awaiting the Proversity entity agreement.

**Comment by gsong** [comment]
@nicovanniekerk I'm going to close this PR for now. Feel free to re-open or open a new PR if/when you're ready to pick this back up.

**Comment by nicovanniekerk** [comment]
@gsong Sorry for letting this hang, i was busy upgrading to gingko the last month. I have sent through the newly signed entity agreement to legal and attaching it here.

Please let me knwo if it is all good now and  will reopen the pull requet.

thanks 
agreement1
agreement2

**Comment by nicovanniekerk** [comment]
@gsong can you please reopen this pull request, we have sorted the agreements with legal.
Please let me know if there is something I should do here, like updating with master and documenting and so on.

thanks

**Comment by nedbat** [comment]
@nicovanniekerk feel free to reopen the pull request.  We have your legal agreement sorted out.

**Comment by nedbat** [comment]
Oops, sorry, my bad.

**Comment by nicovanniekerk** [comment]
@nedbat Thanks for re-opening, please let me know if there is still something needed from my side.

File: playbooks/roles/edx_ansible/templates/update.j2 (hunk 2 above)
Comment (nedbat) [reply playbooks/roles/edx_ansible/templates/update.j2]: @nicovanniekerk is this line right?

Comment (nedbat) [reply playbooks/roles/edx_ansible/templates/update.j2]: Also, ideally, these lines would be indented four spaces like the rest of the file.

Comment (nicovanniekerk) [reply playbooks/roles/edx_ansible/templates/update.j2]: @nedbat I have fixed the spacing, sorry about that, and yes the sed command changes the content of opt variable, the same as Ruby handles the "!" methods.

**Comment by nicovanniekerk** [comment]
@nedbat @feanil Please let me know if there is still something to do here. thanks

File: playbooks/roles/edx_ansible/templates/update.j2 [hunk 3]
```diff
@@ -44,6 +48,12 @@ fi
 extra_args="$extra_args -e 'disable_edx_services=true' -e devstack=true"
 {% endif %}
 
+if [[ "$3" == "-e" ]]; then
+    opt=$4
+    sed -e 's/^"//' -e 's/"$//' <<<"$opt"
```
Comment (feanil) [inline playbooks/roles/edx_ansible/templates/update.j2:53]: @nicovanniekerk how does this store the new value back in the var.  Doesn't this just process `$opt` through sed but then drop the output on the ground?

Comment (nicovanniekerk) [reply playbooks/roles/edx_ansible/templates/update.j2:53]: @feanil It does magic :) the sed command changes what ever is fed too it, so it handles the content of the opt variable and the changes stays in the opt variable, so on the next command the opt variable is changed.

from the sed man page:
Sed is a stream editor. A stream editor is used to perform basic text transformations on an input stream (a file or input from a pipeline). While in some ways similar to an editor which permits scripted edits (such as ed), sed works by making only one pass over the input(s), and is consequently more efficient. But it is sed's ability to filter text in a pipeline which particularly distinguishes it from other types of editors

**Comment by nicovanniekerk** [comment]
@feanil @nedbat anything neeed to be done here, can you merge it please?

**Comment by nedbat** [comment]
@edx/devops any progress?

**Comment by feanil** [comment]
I'm testing this now, assuming that goes well, we can close this out some time this week.

**Review by feanil — APPROVED** [review]
I tested this and it looks like it works as expected.

**Comment by openedx-webhooks** [comment]
@nicovanniekerk 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/configuration&entry.867055334=https://github.com/edx/configuration/pull/3929&entry.1484655318=https://github.com/nicovanniekerk&entry.752974735=2017-06-20+12:57&entry.1917517419=2018-03-20+20:49&entry.2133058324=Yes) so we can improve your experience in the future.

**Pull request merged**